opennms-base-assembly/src/main/filtered/etc/examples/script-policies/example.groovy

```diff
@@ -26,29 +26,103 @@
+import org.opennms.netmgt.model.OnmsAssetRecord
+import org.opennms.netmgt.model.OnmsIpInterface
+import org.opennms.netmgt.model.OnmsSnmpInterface
+import org.opennms.netmgt.model.PrimaryType
 
-for(OnmsIpInterface iface : node.getIpInterfaces()) {
+import java.util.stream.Collectors
 
-    // In this case the 192.168.100.0/24 interface should be primary while the remaining should be set to SECONDARY
+// transaction is a boolean to indicate whether the scripts run in transaction or not.
+// Not in transaction indicates that the node is in initial stage of scanning phase and  may not be persisted yet.
+// Only use this to add assets/metadata or return invalid node (returning null) to abort the scan.
 
-    if (iface.getIpAddressAsString().matches("^192\\.168\\.100\\..*")) {
-        LOG.warn(iface.getIpAddressAsString() + " set to PRIMARY")
-        iface.setIsSnmpPrimary(PrimaryType.PRIMARY)
-    } else {
-        LOG.warn(iface.getIpAddressAsString() + " set to SECONDARY")
-        iface.setIsSnmpPrimary(PrimaryType.SECONDARY)
+if(!transaction) {
+    LOG.debug("customscript: not in transaction");
+    OnmsAssetRecord assetRecord = new OnmsAssetRecord();
+    assetRecord.setAssetNumber("12345");
+    node.setAssetRecord(assetRecord);
+    return node;
+}
+
+// Modifying primary interface should be done in transaction mode.
+
+// Example script for modifying the primary interface
+
+List<String> interfaceNames = Arrays.asList("Lo0","lo0","lo0.0","loopback0","Lo222","Lo201"
+        ,"mgt0","mgmt0","mgmt","primary","bond0","eth0","Vl103","management");
+
+List<OnmsIpInterface> interfacesWithSNMP = node.getInterfacesWithService("SNMP");
+
+LOG.debug("customscript : Interfaces with SNMP service {}", interfacesWithSNMP);
+
+List<OnmsSnmpInterface> snmpInterfaces = interfacesWithSNMP.stream()
+        .filter({iface -> iface.getSnmpInterface() != null})
+        .map({iface -> iface.getSnmpInterface()}).collect(Collectors.toList());
+
+List<String> ifNames = snmpInterfaces.stream()
+        .filter({snmpInterface -> snmpInterface.getIfName() != null})
+        .map({snmpInterface -> snmpInterface.getIfName()}).collect(Collectors.toList());
+
+LOG.debug("customscript : ifNames  {}", ifNames);
+
+String matchedIfName = interfaceNames.stream()
+        .filter({ interfaceName -> ifNames.contains(interfaceName) })
+        .findFirst()
+        .orElse(null);
+
+OnmsIpInterface ipInterface = null;
+
+if (matchedIfName == null) {
+    // Find lowest IP Address as there is no match in if names.
+    List<byte[]> ipAddressesInBytes = interfacesWithSNMP.stream().map({ onmsIpInterface -> onmsIpInterface.getIpAddress().getAddress() }).collect(Collectors.toList());
+    byte[] lowestNumberedIp = Collections.min(ipAddressesInBytes, new ByteArrayComparator());
+    ipInterface = interfacesWithSNMP.stream()
+            .filter({ onmsIpInterface -> Arrays.equals(onmsIpInterface.getIpAddress().getAddress(), lowestNumberedIp) })
```

Comment:
We already ran through the list above, and now we're going through it again to find the match. Can we avoid visiting each IP interface twice?